fix(app-shell): detect usb devices after update

sfoster1 opened the pull request:
When we updated to electron 27, our old package for detecting devices didn't work anymore. Instead we now use the npm package `usb`, which is a reimplementation of node-usb that also provides a WebUSB frontend.

The integration here had a couple small problems; the webusb just straight up doesn't seem to work with these devices, and we were doing the callback glue slightly wrong.

With both those problems fixed, we should once again detected device add and removal basically instantly, which fixes some issues with robot detection and removal.

Closes RQA-2324

sfoster1 committed:
fix(app-shell): detect usb devices after update

When we updated to electron 27, our old package for detecting devices
didn't work anymore. Instead we now use the npm package `usb`, which is
a reimplementation of node-usb that also provides a WebUSB frontend.

The integration here had a couple small problems; the webusb just straight
up doesn't seem to work with these devices, and we were doing the
callback glue slightly wrong.

With both those problems fixed, we should once again detected device add
and removal basically instantly, which fixes some issues with robot
detection and removal.

Closes RQA-2324

sfoster1 committed:
fix tests

brenthagen commented on app-shell/src/system-info/usb-devices.ts:
```diff
@@ -12,30 +12,142 @@ export type UsbDeviceMonitorOptions = Partial<{
+      return descriptorToDevice(device, manufacturer, serialNumber, productName)
+    })
+    .finally(() => {
+      device.close()
```
i'm getting `[shell] 15:29:48.027 [usb-devices] error: Failed to format added device vid=5ac pid=5ac for upstream: Error: Can't close device with a pending request` from this close call on device add, trying to see why. I'm assuming we need to close the device here?

brenthagen replied:
putting the close call on a 1 sec timeout seems to work but...why is it necessary

sfoster1 replied:
I assume so yeah, definitely a bug in how we call these promises

sfoster1 committed:
allow individual descriptor accesses to fail

sfoster1 committed:
display-side unknown fallbacks

sfoster1 committed:
format

sfoster1 committed:
some smaller fixes

sfoster1 committed:
give usb devices identifiers

I think previously the library cached this for us, but now when a device
is removed we don't actually get any of the string descriptor
information. That includes serial. That means that the comparison that
lets us remove devices from redux is wrong.

Either we could cache the serial on the node side, which seems annoying,
or we could use the unique identifiers that libusb gives us, which is
bus+device address. Device addresses must be unique per device per bus.

sfoster1 committed:
fix some logging stuff

sfoster1 committed:
tolerate partial string desc fetches

sfoster1 commented:
Couple tests left to pass but @brenthagen i think that issue is fixed. right now the only problem appears to be that we can't reliably pull mfr and product name on windows at least for composites

sfoster1 commented:
i'm worried that might be this: https://github.com/libusb/libusb/commit/fdab67b14ac32891158dc210636b622c42e24e22 which is fixed in libusb 0.27.0 which is of course one later version than what node-usb vendors

sfoster1 committed:
implement windows using powershell

Oh boy

brenthagen reviewed (approved):
talked through with @sfoster1 , checked device add/remove at least on the macOS side without issue (including Flex)

sfoster1 committed:
test fixups including not testing windows

sfoster1 committed:
eslint
